### PR TITLE
delay chain validation

### PR DESCRIPTION
**Pull request by witmicko**
## Explanation

Fixing https://github.com/MetaMask/pm-security/issues/120
An arbitrary HTTP POST request is issued on wallet_addEthereumChain on any website. This happens even on not connected websites (e.g. [github.com](https://github.com/) ), even if the wallet is locked. It does not happen only if there is no wallet.

Fixed by making confirmation approve handler asynchronous.

### After

https://user-images.githubusercontent.com/3279642/219458534-086b49b5-37d2-4fa7-b57c-a36751529500.mov

## Pre-merge author checklist

- [ ] I've clearly explained:
  - [ ] What problem this PR is solving
  - [ ] How this problem was solved
  - [ ] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (Gudahtt) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:282]: If either this check or the subsequent one fails, we would want to remove the custom RPC endpoint from the user's wallet because it failed validation.

We should also probably prevent the user from using this network while this validation is ongoing.

Comment (Gudahtt) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:267]: Update: we talked and decided delaying adding the chain was the easier path. This resolves these concerns.

**Review by NEllusion — DISMISSED** [review]
Code looks good, and tested that the fix solves the issue 🎉

File: test/e2e/tests/chain-interactions.spec.js
Comment (Gudahtt) [inline test/e2e/tests/chain-interactions.spec.js:53]: Why was this additional pause and request for window handles required? I can see that the webdriver was already pointing at the notification popup, but it's not clear to me what happens here.

Is the popup closing and re-opening again?

Comment (witmicko) [reply test/e2e/tests/chain-interactions.spec.js:53]: window id is changing, approve add network has different ID than confirm switch network. Added pause to catch this change, witouth these changes it was failing with `window not found`

Comment (Gudahtt) [reply test/e2e/tests/chain-interactions.spec.js:53]: Ok, that implies that the window is closing and re-opening.

I'm having trouble understanding how your changes relate to this though. You didn't adjust the switch network steps at all.

Perhaps some part of your changes is causing the window to close sooner than it had before? Ideally it would stay open long enough for the switch network prompt to appear.

Comment (witmicko) [reply test/e2e/tests/chain-interactions.spec.js:53]: My change makes it stay open longer I think as I moved validation and adding network to after user approves. Maybe that is why window was missing in test, because previously new window would have been already opened by then?
Let me know if you would like me to make some changes please.

Comment (legobeat) [reply test/e2e/tests/chain-interactions.spec.js:53]: I don't have any insight to share on the actual issue, just noting that it seems optimistic with a 1s window here. Starvation scenarios are not that rare.

Comment (Gudahtt) [reply test/e2e/tests/chain-interactions.spec.js:53]: We should definitely try to prevent the window from closing. This would be a substantial UX degradation, and would make these e2e tests flaky.

**Review by brad-decker — COMMENTED** [review]
This looks good to me from a code perspective. i'll QA now

**Comment by brad-decker** [comment]
We should probably get some UX feedback on how we display the loading. Right now this is leading to two lines of text on the button and causing a little bit of layout jank. See the video.

@bschorchit do you know of what other confirmations do when loading (if anything)? 

[Screencast from 02-02-2023 12:10:53 PM.webm](https://user-images.githubusercontent.com/4448075/216409952-10694ce2-a415-46fd-8088-62c0d2e9526e.webm)

**Comment by bschorchit** [comment]
> @bschorchit do you know of what other confirmations do when loading (if anything)?
> 
Oh good question, we usually use the spinner on the full screen but that's when loading the UI. I don't think I've seen a loading after submitting yet. Any insights or suggestion on this @holantonela ?

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (Gudahtt) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:370]: Nit: Perhaps better to return `undefined` if the return value is not expected to be used

```suggestion
      return undefined;
```

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (Gudahtt) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:356]: This is a clever solution to the double-window-popup problem :+1: Great work

The one major flaw left is that this experience has been made more confusing for the user because of this change, specifically in the case where this validation fails.

Previously if this validation step failed, the dapp would get an error and the user would see nothing. In this circumstance, maybe the dapp handled it well and maybe they didn't, but at least MetaMask isn't doing anything to mislead/confuse users.

But now if validation fails, it will appear to a user that the operation succeeded. Only the dapp and the web console are told of the error. The user may not find out.

It would be better to show the user the error somehow. But I don't know of an easy way to do this :thinking:

Comment (Gudahtt) [reply ui/pages/confirmation/templates/add-ethereum-chain.js:356]: Maybe this UX concern is not a blocker, and we can solve it in a later PR. We should at least track this problem as an issue if it's not solved here though.

Comment (NicholasEllul) [reply ui/pages/confirmation/templates/add-ethereum-chain.js:356]: As a temporary workaround until the other work to support this is complete, can we infer if the operation failed by following up with a second check to see what chain ID we are using?

Similar to whats done in:

https://github.com/MetaMask/metamask-extension/blob/cdc1bce6884295e97c63ee9758646fc4287aa781/ui/components/app/network-display/network-display.js#L31-L34

Comment (NicholasEllul) [reply ui/pages/confirmation/templates/add-ethereum-chain.js:356]: `state.metamask.provider` gives us access to the following to check the chain ID: 


Screenshot 2023-02-15 at 4 34 24 PM

File: app/_locales/en/messages.json
Comment (NicholasEllul) [inline app/_locales/en/messages.json:2028]: Should this read

`Chain ID returned by the RPC URL`

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (NicholasEllul) [reply ui/pages/confirmation/templates/add-ethereum-chain.js]: Update: Michał and I discussed offline and he has found another solution to this problem

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (NicholasEllul) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:373]: Do we still want to be rejecting the pending approval? Or do we no longer need this now that the user is forced in to clicking cancel themselves

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (NicholasEllul) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:376]: This function now returns one of the following depending on if the function succeeds or fails:

* Above we return `actions.rejectPendingApproval` during a failure which is a [thunk object](https://github.com/MetaMask/metamask-extension/blob/33cc8d587aad05c0b41871ba3676676a3ce5680e/ui/store/actions.ts#L3731). (Note: I don't think this value is actually used anywhere)
* Here we return a javascript object containing the `errors` key with an array of strings
* Below we return a javascript object containing an empty`{}`

Is there a way we can consolidate these to provide a consistent return interface? Checking usage of the code, it seems the return value from this function hasn't used in the past, so we have some flexibility to set things up.

File: ui/pages/confirmation/confirmation.js
Comment (danjm) [inline ui/pages/confirmation/confirmation.js:44]: `rekmarks` is actually a github username

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (danjm) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:185]: why is this section of code removed?

Comment (witmicko) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:185]: > why is this section of code removed?

these checks were moved to the confirmation template

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (Gudahtt) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:258]: Missing an `await` here

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:268]: Hmm, was this extra `await` needed? This is already being `await`-ed three lines up from here

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:274]: It looks like this component will be stuck in the "loading" state if the submit result is missing this property. Yet only `add-ethereum-chain` returns this. The other confirmations would all get stuck.

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:384]: There appears to be no way to dismiss these alerts. The `dismissAlert` function here is not for the "submit alerts" (which are component state), it's for the "alert state" reducer. When this line is called, it won't work because the reducer isn't tracking this id.

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (Gudahtt) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:369]: Nit: we should at least confirm that this is a network error, and throw otherwise. That would help ensure that we don't swallow errors in our code.

File: ui/pages/confirmation/confirmation.js
Comment (witmicko) [reply ui/pages/confirmation/confirmation.js:274]: this is done this way to disable the button so user can only click 'Cancel'

Comment (Gudahtt) [reply ui/pages/confirmation/confirmation.js:274]: Right, but for the non-`add-ethereum-chain` confirmations there will be no way to proceed even if there is no error

File: app/_locales/en/messages.json
Comment (NicholasEllul) [inline app/_locales/en/messages.json:1360]: Can we use the terminology `Network` here and in the translation above to align with the terminology used for the other UI elements?

**Review by NEllusion — COMMENTED** [review]
Tested with reproduction steps listed in the GitHub issue and the change works as expected

Minor comment: We now end up showing `Adding RPC...` after the request fails even when the process of adding an RPC is terminated. Is there a way we can fall back to showing `approve` in a disabled state? Alternatively, could we get away with simply re-activating the button and having the error show up again?

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:280]: It looks like `submitResult.success` is no longer being used. Perhaps we could simplify this to just return the error, and drop the `success` boolean flag altogether?

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:267]: Nit: Sorry I know I requested that the inner `await` be removed earlier. But I hadn't considered which we needed here, I just noticed the redundancy. On second look, it does seem to make more sense for the `await` to just be used for `onSubmit`, since `useWarningModal` is a boolean property not a Promise

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:276]: We're missing this "submit result error handling" step for confirmations that include a warning modal. For the warning modal case, `onSubmit` is called in the `onSubmit` prop passed to `ConfirmationWarningModal` below.

Perhaps we can move this result handling to a separate function/wrapper function, and use it in both places?

File: ui/pages/confirmation/templates/index.js
Comment (Gudahtt) [reply ui/pages/confirmation/templates/index.js]: Sorry for the confusion; I meant this to apply just to `'submitAlerts'`. The `'loadingText'` property is being used in the `add-ethereum-chain` template.

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [reply ui/pages/confirmation/confirmation.js:277]: Could you remove this outer `await`, now that the inner one has been restored?

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (Gudahtt) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:374]: Hmm. I'm not confident that all fetch errors would have this exact message. It looks like they can have a variety of different messages, especially if we consider that custom RPC endpoints can return custom errors for auth failures. I guess my suggestion for identifying the error as being a "connectivity" is impractical for now.

File: ui/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:285]: It seems that this is getting called even in the `setShowWarningModal` case. In that case, `submitResult` would be `undefined`, which results in `setLoading(false)` being called. This effectively breaks the loading state for confirmations with warnings (at least I think so - I didn't test this.

It might be simpler to change this ternary into if-else blocks, so that we can ensure this is only called when `onSubmit` is called.

File: test/e2e/tests/add-custom-network.spec.js
Comment (Gudahtt) [inline test/e2e/tests/add-custom-network.spec.js:70]: Nit: Is this delay needed? The `findElement` call will automatically wait until it finds the requested element, we don't usually need these explicit delays.

```suggestion
```

File: test/e2e/tests/add-custom-network.spec.js
Comment (Gudahtt) [inline test/e2e/tests/add-custom-network.spec.js:91]: Nit: These assertions are redundant; the `findElement` call will throw unless it finds something with that exact text. It's already asserting this for us. Removing these assertions saves another round-trip communication with the browser, reducing test runtime and flakiness.

```suggestion
        await driver.findElement({
          tag: 'h4',
          text: warningTxt,
        });

        await driver.clickElement({
          tag: 'button',
          text: 'Approve',
        });

        const errMsg =
          'Chain ID returned by the custom network does not match the submitted chain ID.';
        await driver.findElement({
          tag: 'span',
          text: errMsg,
        });
```

Comment (Gudahtt) [reply test/e2e/tests/add-custom-network.spec.js:91]: I see there are a few examples of this in the other tests as well

File: test/e2e/tests/add-custom-network.spec.js
Comment (Gudahtt) [inline test/e2e/tests/add-custom-network.spec.js:58]: Nit: Do we need to double the timeout here? 5 seconds is the default; I would expect the notification to appear in that timeframe.
```suggestion
        const windowHandles = await driver.waitUntilXWindowHandles(3);
```

File: test/e2e/tests/add-custom-network.spec.js
Comment (Gudahtt) [inline test/e2e/tests/add-custom-network.spec.js:21]: Maybe this title should be updated to reflect that the test is for the scenario where a warning is shown first, followed by an error. It's good that we have this case covered, but I hadn't expected that from this description.

File: ui/pages/confirmation/templates/index.js
Comment (Gudahtt) [reply ui/pages/confirmation/templates/index.js]: @witmicko Looks like this is still unaddressed; this is the problem I referred to on Zoom.

It's not breaking the e2e test because the validation is broken (pre-existing issue, I'll address it now in a separate PR)

File: test/e2e/tests/add-custom-network.spec.js
Comment (Gudahtt) [inline test/e2e/tests/add-custom-network.spec.js:136]: Nit: Shouldn't need double timeout here

```suggestion
        const windowHandles = await driver.waitUntilXWindowHandles(3);
```

File: test/e2e/tests/add-custom-network.spec.js
Comment (Gudahtt) [inline test/e2e/tests/add-custom-network.spec.js:205]: Nit: Shouldn't need double timeout here
```suggestion
        const windowHandles = await driver.waitUntilXWindowHandles(3);
```

**Review by Gudahtt — COMMENTED** [review]
Great work! Last blocking issue is to add `loadingText` as an allowed return value of `getValues` ([see here](https://github.com/MetaMask/metamask-extension/pull/17413#discussion_r1126741375))

**Review by Gudahtt — COMMENTED** [review]
LGTM! Though now there's a conflict with the coverage target file

**Review by Gudahtt — DISMISSED** [review]

**Review by Gudahtt — APPROVED** [review]

**Comment by danjm** [comment]
Created three unit test creation issues for either the extension platform team or the confirmations team to pick up:
https://github.com/MetaMask/metamask-extension/issues/18060
https://github.com/MetaMask/metamask-extension/issues/18059
https://github.com/MetaMask/metamask-extension/issues/18058

**Review by danjm — APPROVED** [review]
LGTM! Nice work on the e2e tests

**Pull request merged**
